### ISSUE TITLE
TreeNode memory leak and multiple getChildren calls

### ISSUE THREAD

**Issue by stogle** [500tech/angular-tree-component#582]
**Steps to reproduce**

This issue can be reproduced with release 7.0.1 of this GitHub repository.

1. Run `npm run example:cli:win`
2. In Chrome, open http://localhost:4200
3. Open the browser console.
4. Expand "asyncroot". Observe that the console shows one "toggleExpanded" event and one "loadNodeChildren" event, which is expected.
5. (Important) Refresh the browser to clear the asynchronously loaded children.
6. Click the "Add Node" button **ten** times.
7. (Optional) Use the browser developer tools to try to force garbage collection.
7. Expand "asyncroot". Observe that the console shows one "toggleExpanded" event and **eleven** "loadNodeChildren" event.

**Observations**

1. Every time `TreeModel.update()` is called a new `TreeNode` object is created for every node in the tree. The `TreeNode` objects never seem to be destroyed, resulting in a significant memory leak.
2. On expanding a node that loads children asynchronously, **all** the `TreeNode` objects that have ever been created for that node will call `getChildren`. If `getChildren` makes a HTTP request this will result in multiple identical requests being issued in rapid succession.

This makes the component virtually unusable in any scenario that involves updating a tree with lazy loading.

**Comment by stogle** [comment]
My theory on the cause of this issue is that the reaction in [`TreeNode.autoLoadChildren`](https://github.com/500tech/angular-tree-component/blob/a3d58cdce63c0fa24b0ce1015f892cdbb94ce83d/lib/models/tree-node.model.ts#L276) is never disposed (https://mobx.js.org/best/pitfalls.html#always-dispose-reactions). The class should have some sort of `dispose` method that disposes of this reaction and then calls `dispose` on each of its children. This `dispose` method should be called on the virtual root in [`TreeModel.update`](https://github.com/500tech/angular-tree-component/blob/69a3200719535a53d4c4e2209e29b0e0b9139c26/lib/models/tree.model.ts#L177) before replacing it.

**Comment by adamkleingit** [comment]
@stogle - can you check in 7.1.0? The disposing of reaction is published there
Thanks

**Comment by KBRON111** [comment]
@adamkleingit @stogle I can still reproduce the memory leak in the example with the latest build. I nulled out the parent and children of the node in the [TreeNode.Dipose](https://github.com/500tech/angular-tree-component/blob/3b3caa749e5c9a4a584a36e1e2ad7d7063ab334d/lib/models/tree-node.model.ts#L297) method:

```
  dispose() {
    if (this.children) {
      this.children.forEach((child) => child.dispose());
    }
    if (this.handler) {
      this.handler();
    }
    this.children = null;
    this.parent = null;
  }
```

and the leak improves greatly, but there is still some (relatively) minor leaking happening with ever update even after those changes.

**Comment by adamkleingit** [comment]
@KBRON111 can you make a PR?

**Comment by kareljuricka** [comment]
Great work with memory leak fixes.
It's possible to release new version from this updated master?

**Comment by adamkleingit** [comment]
Published in 8.0.1.

Thanks for the PR!
